### PR TITLE
[ENT-125] Dynamically load course modes from enrollment API

### PR DESCRIPTION
**Pull request by haikuginger**
**Description:**

Previously, the course mode box for enrolling users from the Enterprise app was limited to three "blessed" modes - audit, verified, and professional. However, it became clear that this was insufficient.

This PR modifies the view to dynamically load all available enrollment modes for a course directly from the Enrollment API, and presents them with the custom names given to them, rather than label them by internal value.

In addition, the other three valid modes (no-id-professional, credit, and honor) have been added to the allowable options, meaning that in cases where Javascript is disabled, all valid modes can be chosen correctly.

**Screenshots:**

No course ID entered:

screen shot 2017-01-04 at 9 22 26 pm

Valid course ID entered:

screen shot 2017-01-04 at 9 23 42 pm

Invalid course ID entered:

screen shot 2017-01-04 at 9 24 05 pm

Javascript disabled:

screen shot 2017-01-04 at 9 25 52 pm

**JIRA:** https://openedx.atlassian.net/browse/ENT-125

**Dependencies:** None

**Merge deadline:** 2017-01-17

**Sandbox**: https://ent91.opencraft.hosting (built-in `staff` user)

**Testing instructions:**

1. Create a couple custom course modes for a given course; use distinctive names.
2. Navigate to the Manage Learners view of the Enterprise Admin.
    * Observe that the course mode dropdown is disabled, and is filled with a prompt to enter a valid course ID.
2. Enter the course ID of the course you've chosen and tab away from the course box.
    * Observe that the course mode dropdown is populated with the course modes you created.
3. Select one of the modes and submit the form.
    * Observe that although the form has errors, the course mode dropdown is returned to the same state and position it was in prior to submitting the form.
4. Fill in the email address of an existing user not yet linked with this EnterpriseCustomer, and submit the form again.
    * Observe that the user is now enrolled in the given course, with the given course mode you chose from the dropdown.
3. Enter an invalid course ID in the course ID box.
    * Observe that the course mode box returns to its original, disabled, state.
4. Disable Javascript and refresh the page.
    * Observe that although nothing is dynamically loaded, all six valid course mode options are presented for you to select from.

**Reviewers:**
- [x] @e-kolpakov  
- [x] EdX TBD

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:**

1. I have some concerns about a11y with this PR. Multiple browsers (I tested Safari and Firefox) appear to choose tab order prior to calling the `.blur()` event on the control that's being departed. As a result, if the mode drop-down is disabled prior to tabbing away from the course ID box, then the focus will go to the submit button, rather than the dropdown, even if the AJAX call will enable the dropdown. There are a couple ways around this; none pretty:
    * Whenever course mode items are successfully loaded, set the focus to the dropdown. This is ugly because it prevents the user from choosing whatever focus item they like.
    * Don't actually disable the dropdown when there aren't valid options. This isn't ideal, but doesn't cause any harm, because when it's in such a state, there's only one option to choose from anyway.

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=26&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by e-kolpakov** [comment]
@haikuginger All looks good, except for a couple of things:

1. Version bump?
2. Changelog record?
3. (most important) That little JS file starts to grow - and there are no tests for it at all. Might be a good time to introduce some JS testing framework (i.e. Jasmine, qUnit, or anything you're familiar with).

**Comment by mattdrayer** [comment]
@asadiqbal08 @saleem-latif @zubair-arbi  FYI on this PR --

**Comment by haikuginger** [comment]
Rebased from f0cee6a.

@e-kolpakov, that 500 error should be fixed - an internal API client was trying to make an HTTPs connection to the LMS, which the hosts file listed as being at 127.0.0.1. Since we terminate HTTPs on the load balancer, this resulted in a failure of the API client to connect. I've also bumped the version and added a preliminary changelog entry.

File: enterprise/static/enterprise/admin/manage_learners.js [hunk 1]
```diff
@@ -1,5 +1,42 @@
...
     function () {
+        function disableMode(){
+            $("#id_course_mode").each(function(index, dropdown){
+                dropdown.options.length = 0
+                dropdown.options[0] = new Option(gettext('Enter a valid course ID'), '')
+                dropdown.disabled = true
+            });
+        }
+        function loadCourseModes(courseId){
+            $("#id_course").each(function(i, entry){
```
Comment (asadiqbal08) [inline enterprise/static/enterprise/admin/manage_learners.js]: `i` ~ `index` ?

File: enterprise/static/enterprise/admin/manage_learners.js [hunk 2]
```diff
@@ -1,5 +1,42 @@
...
+                courseId = entry.value
+                if (courseId === ''){
+                    disableMode()
```
Comment (asadiqbal08) [inline enterprise/static/enterprise/admin/manage_learners.js]: nit: disableMode();

File: enterprise/static/enterprise/admin/manage_learners.js [hunk 3]
```diff
@@ -18,5 +55,20 @@ $(
                 })
             }
         });
+        $("#id_course").on('input', function(evt){
```
Comment (asadiqbal08) [inline enterprise/static/enterprise/admin/manage_learners.js]: space around open bracket. 
nit: `evt` ~ `event`

File: enterprise/static/enterprise/admin/manage_learners.js [hunk 4]
```diff
@@ -18,5 +55,20 @@ $(
...
+            /*
+            We have to do two-step input/blur because some browsers (looking at you, Safari)
+            don't properly send the .change() event when the input is filled with autocomplete.
+            */
+            evt.target.dirty = true;
+        });
+        $("#id_course").blur(function(evt){
```
Comment (asadiqbal08) [inline enterprise/static/enterprise/admin/manage_learners.js]: space around open bracket. many places above . 
nit: good to have : `evt` ~ `event`

File: enterprise/static/enterprise/admin/manage_learners.js [hunk 5]
```diff
@@ -18,5 +55,20 @@ $(
...
+            var target = evt.target;
+            if (target.dirty){
```
Comment (asadiqbal08) [inline enterprise/static/enterprise/admin/manage_learners.js]: space around open bracket. 

e.g. 
`if (target.dirty) {`

File: enterprise/static/enterprise/admin/manage_learners.js [hunk 6]
```diff
@@ -18,5 +55,20 @@ $(
...
+                loadCourseModes();
+                target.dirty = false;
+            }
+        });
+        loadCourseModes()
```
Comment (asadiqbal08) [inline enterprise/static/enterprise/admin/manage_learners.js]: what is the purpose of double calling `loadCourseModes()` ? 
nit: loadCourseModes()`;`

File: enterprise/static/enterprise/admin/manage_learners.js (hunk 3 above)
Comment (e-kolpakov) [inline enterprise/static/enterprise/admin/manage_learners.js]: @asadiqbal08 @haikuginger technically it should be "space before opening curly-brace". :)

File: enterprise/static/enterprise/admin/manage_learners.js (hunk 6 above)
Comment (haikuginger) [inline enterprise/static/enterprise/admin/manage_learners.js]: It isn't being double-called. It's executed once on document ready, and as part of `$("#id_course").blur()`.

**Comment by haikuginger** [comment]
@asadiqbal08, your comments have been addressed.

**Comment by haikuginger** [comment]
Apologies to all for the ongoing emails as I try to get Travis to play nice with Jasmine.

**Comment by haikuginger** [comment]
@e-kolpakov, JS test suite added.

File: spec/javascripts/enterprise_admin_spec.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                        name: 'blah',
+                        slug: 'really squishy'
+                    },
+                    {
+                        name: 'booooo',
+                        slug: 'spooky ghost'
+                    }
+                ]
+            });
+            done()
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: done()`;`

File: spec/javascripts/enterprise_admin_spec.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        beforeEach(function(done) {
+            result = fillModeDropdown({
+                course_modes: [
+                    {
+                        name: 'blah',
+                        slug: 'really squishy'
+                    },
+                    {
+                        name: 'booooo',
+                        slug: 'spooky ghost'
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: can we use some proper wording here too ?

File: spec/javascripts/enterprise_admin_spec.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    describe("when filling the mode dropdown with data", function(){
+        beforeEach(function(done) {
+            result = fillModeDropdown({
+                course_modes: [
+                    {
+                        name: 'blah',
+                        slug: 'really squishy'
+                    },
+                    {
+                        name: 'booooo',
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: can we use some proper wording here instead of `booooo` ?

File: spec/javascripts/enterprise_admin_spec.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        });
+    });
+
+    describe("when filling the mode dropdown with data", function(){
+        beforeEach(function(done) {
+            result = fillModeDropdown({
+                course_modes: [
+                    {
+                        name: 'blah',
+                        slug: 'really squishy'
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: can we use some proper wording here too ?

File: spec/javascripts/enterprise_admin_spec.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+            expect($("#id_course_mode")).toBeDisabled();
+        });
+    });
+
+    describe("when filling the mode dropdown with data", function(){
+        beforeEach(function(done) {
+            result = fillModeDropdown({
+                course_modes: [
+                    {
+                        name: 'blah',
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: can we use some proper wording here instead of `blah` ?

File: spec/javascripts/enterprise_admin_spec.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        jasmine.Ajax.install();
+    });
+
+    afterEach(function(){
+        jasmine.Ajax.uninstall();
+    });
+
+    describe("on page load", function(){
+        it("indicates that a valid course ID is required on page load.", function(){
+            loadPage();
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: instead of calling `loadPage();` in each function explicitly, you can put one place -- `beforeEach`

File: spec/javascripts/enterprise_admin_spec.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
+var enrollmentApiRoot = 'http://localhost/'
+var gettext = function(arg){
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: usually we are following to add space before opening braces in javascript files. better to follow the convention in all over places below too.

File: enterprise/static/enterprise/admin/manage_learners.js [hunk 14]
```diff
@@ -1,22 +1,80 @@
...
-                })
+function disableMode() {
+    $("#id_course_mode").each(function (index, dropdown) {
+        dropdown.options.length = 0;
+        dropdown.options[0] = new Option(gettext('Enter a valid course ID'), '');
+        dropdown.disabled = true;
+    });
+}
+
+function fillModeDropdown(data) {
```
Comment (asadiqbal08) [inline enterprise/static/enterprise/admin/manage_learners.js]: can you add description that represent what this function `fillModeDropdown` is actually doing.

File: spec/javascripts/enterprise_admin_spec.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            expect(dropdown).toContainElement("option[value='spooky ghost']")
+            expect(dropdown).toHaveText("blahbooooo")
+        });
+
+    });
+
+    it("makes the correct API call for course mode", function(){
+        loadPage();
+        $("#id_course").val("course1")
+        loadCourseModes()
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: loadCourseModes()`;`

File: spec/javascripts/enterprise_admin_spec.js [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+            expect(dropdown).toContainElement("option[value='really squishy']")
+            expect(dropdown).toContainElement("option[value='spooky ghost']")
+            expect(dropdown).toHaveText("blahbooooo")
+        });
+
+    });
+
+    it("makes the correct API call for course mode", function(){
+        loadPage();
+        $("#id_course").val("course1")
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: `;`

File: spec/javascripts/fixtures/learner_admin.html [hunk 17]
```diff
@@ -0,0 +1,6 @@
+<select id="id_course_mode">
+    <option value="real">Real value</option>
+    <option value="fake">Fake value</option>
+</select>
+
+<input id="id_course" value="">
```
Comment (asadiqbal08) [inline spec/javascripts/fixtures/learner_admin.html]: line missing ?

File: spec/javascripts/enterprise_admin_spec.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        }, 5000)
+
+        it("renders correctly", function() {
+            dropdown = $("#id_course_mode")
+            expect(dropdown).not.toBeDisabled()
+            expect(dropdown).toContainElement("option[value='really squishy']")
+            expect(dropdown).toContainElement("option[value='spooky ghost']")
+            expect(dropdown).toHaveText("blahbooooo")
+        });
+
```
Comment (asadiqbal08) [inline spec/javascripts/enterprise_admin_spec.js]: extra line

File: spec/javascripts/fixtures/learner_admin.html [hunk 19]
```diff
@@ -0,0 +1,6 @@
+<select id="id_course_mode">
```
Comment (e-kolpakov) [inline spec/javascripts/fixtures/learner_admin.html]: @haikuginger That's not even 10% of learner_admin HTML :) Suggestion: either rename it (to keep it a minimal amount of markup for course_mode tests) or extend to have a more or less actual HTML.

**Comment by haikuginger** [comment]
@asadiqbal08, @e-kolpakov, updated to meet your requirements. Apologies about the consistent issues with Javascript formatting; I'm just not familiar with the style, so I can't "see" those issues quite yet.

**Comment by e-kolpakov** [comment]
@haikuginger Suggestion: don't try to follow the JS convention - enforce it :) There's a bunch of JS linters, you might want to use jshint ([example commit of adding it](https://github.com/open-craft/xblock-group-project-v2/commit/bebb1304f2249e1991e5d86568296a48ddff8827)) or pick whatever you like.

**Comment by haikuginger** [comment]
@mattdrayer, we've now got this running both Jasmine and jshint in CI; what can I do to help this get merged today to meet our sprint commitment?

**Comment by mattdrayer** [comment]
I think we've gone through several rounds of revisions at this point, and I don't see anything important enough to warrant another set of edits.  Jenkins checks are green, so I say let the code flow -- 👍

**Comment by asadiqbal08** [comment]
Thanks @haikuginger 👍
